testdata/p4_16_samples/pna-dpdk-add_on_miss0.py

```diff
@@ -0,0 +1,185 @@
+            grpc_addr = 'localhost:9559'
+        certs_dir = '/usr/share/stratum/certs'
+        root_certificate = certs_dir + '/ca.crt'
+        private_key = certs_dir + '/client.key'
+        certificate_chain = certs_dir + '/client.crt'
+        # Default to insecure
+        ssl_opts = p4rt.SSLOptions(True, root_certificate, certificate_chain,
+                                   private_key)
+        sh.setup(device_id=1,
+                 grpc_addr=grpc_addr,
+                 election_id=(0, 1),
+                 ssl_options=ssl_opts)
+
+    def tearDown(self):
+        logging.info("IdleTimeoutTest.tearDown()")
+        sh.teardown()
+
+    def runTestImpl(self):
+        self.setupCtrlPlane()
+        logger.info("Sending Packet ...")
+        self.sendPacket()
+        logger.info("Verifying Packet ...")
+        self.verifyPackets()
+
+#############################################################
+# Define a few small helper functions that help construct
+# parameters for the table_add() method.
+#############################################################
+
+def entry_count(table_name):
+    te = sh.TableEntry(table_name)
+    n = 0
+    for x in te.read():
+        n = n + 1
+    return n
+
+def init_key_from_read_tableentry(read_te):
+    new_te = sh.TableEntry(read_te.name)
+    # This is only written to work for tables where all key fields are
+    # match_kind exact.
+    for f in read_te.match._fields:
+        new_te.match[f] = '%d' % (int.from_bytes(read_te.match[f].exact.value, 'big'))
+    return new_te
+
+def delete_all_entries(tname):
+    te = sh.TableEntry(tname)
+    for e in te.read():
+        d = init_key_from_read_tableentry(e)
+        d.delete()
+
+def add_ipv4_host_entry_action_send(ipv4_addr_str, port_int):
+    te = sh.TableEntry('ipv4_host')(action='send')
+    te.match['dst_addr'] = '%s' % (ipv4_addr_str)
+    te.action['port'] = '%d' % (port_int)
+    te.insert()
+
+
+class OneEntryTest(AbstractIdleTimeoutTest):
+
+    def sendPacket(self):
+        in_dmac = 'ee:30:ca:9d:1e:00'
+        in_smac = 'ee:cd:00:7e:70:00'
+        ip_src_addr = '1.1.1.1'
+        ip_dst_addr = '2.2.2.2'
+        ig_port = 0
+        sport = 59597
+        dport = 7503
+        pkt_in = tu.simple_tcp_packet(eth_src=in_smac, eth_dst=in_dmac,
+                                      ip_src=ip_src_addr, ip_dst=ip_dst_addr,
+                                      tcp_sport=sport, tcp_dport=dport)
+        
+        # Send in a first packet that should experience a miss on
+        # table ct_tcp_table, cause a new entry to be added by the
+        # data plane with a 30-second expiration time, and be
+        # forwarded with a change to its source MAC address that the
+        # add_on_miss0.p4 program uses to indicate that a miss
+        # occurred.
+        logging.info("Sending packet #1")
+        tu.send_packet(self, ig_port, pkt_in)
+        first_pkt_time = time.time()
+
+    def verifyPackets(self):
```

Comment:
This extra structure to tests that you have added with verifyPackets and sendPackets methods is fine, for tests where that pattern is sufficient.

In general, the fact that PTF tests allow you to write more general flows of tests, e.g. add some entries, send and verify some packets, add/modify/delete some table entries, send and verify more packets, etc. I consider a super-power of PTF tests over STF tests, and we should not discourage people from using it.

Reply:
Note: I will approve this PR regardless of this comment.  I am not suggesting any changes to this PR.  I just don't want people going forward thinking that they have to restrict themselves to writing tests with this pattern.